src/vs/workbench/api/common/extHostTerminalService.ts

```diff
@@ -823,13 +823,17 @@ export abstract class BaseExtHostTerminalService extends Disposable implements I
 		return index;
 	}
 
-	public getEnvironmentVariableCollection(extension: IExtensionDescription): vscode.EnvironmentVariableCollection {
+	public getEnvironmentVariableCollection(extension: IExtensionDescription): vscode.EnvironmentVariableCollection | { getScopedEnvironmentVariableCollection(scope: vscode.EnvironmentVariableScope | undefined): vscode.EnvironmentVariableCollection } {
 		let collection = this._environmentVariableCollections.get(extension.identifier.value);
 		if (!collection) {
 			collection = new EnvironmentVariableCollection();
 			this._setEnvironmentVariableCollection(extension.identifier.value, collection);
 		}
-		return collection;
+		return {
+			...collection.getScopedEnvironmentVariableCollection(undefined), getScopedEnvironmentVariableCollection(scope) {
+				return collection!.getScopedEnvironmentVariableCollection(scope);
+			},
+		};
```

Comment:
I think it would be simpler if you kept EnvironmentVariableCollection (or DefaultEnvironmentVariableCollection) and ScopedEnvironmentVariableCollection separate, instead of trying to keep them all maintained in a single class. So maintain a `Map<extId, Map<scope, ScopedEnvironmentVariableCollection>>`  in `BaseExtHostTerminalService`

Reply:
Perhaps I misunderstand, but that would require to change how we consume it all around the codebase, so I'm not fond of the idea. For eg.

https://github.com/microsoft/vscode/blob/efb49cc271d6eea5634ac395e36e1e4cd108a447/src/vs/platform/terminal/common/environmentVariableCollection.ts#L22

Everywhere it is assumed collections is `Map<extId, UnifiedExtensionCollectionObject>` instead of `Map<extId, Map<scope, ScopedEnvironmentVariableCollection>>`.

Reply:
Let’s keep going with what you’ve done. This is just the way I would have done it from the start but it looks like it would take a bunch of re-work. Let's make sure to have a bunch of tests around merging the scopes collections if they don't already exist.